### DISCUSSION TITLE
Hash of all properties/relations of a model to be used for versioning when caching full objects

### DISCUSSION THREAD

**Discussion by ddorian** [sqlalchemy/sqlalchemy#11245] [Usage Questions] [answered]
Hi,

So I want to cache full orm models (along with relationships) of the python process externally.
And I need a way to "version" the models, use this version in the cache key, since I may add/remove columns/relationships.
I know I can just iterate on Model.__mapper__.columns/relationships. 

This might become complex when using inheritance, self referential relationships, etc etc.

But maybe there's already a way to do it that I didn't find by searching.

Regards,
Dorian

**Comment by zzzeek** [comment, marked as answer]
what does "cache full orm models " mean here?  you want to have a way to recall specific class structures, like classes with particular columns?      Usually "caching" refers to objects you loaded from queries.

ORM models are Python code.   So you can have your various kinds of ORM models simply in Python source files in some kind of versioned directory structure, if you wanted.   Since you want to have relationships and other custom properties I dont see much other way around that.

this overall seems like a **very** odd thing to want to do, so wondering what it is you are **actually** doing, like a versioned RESTful API?

**Reply by ddorian** [reply]
I'm using the caching query on https://github.com/zzzeek/sqlalchemy/blob/main/examples/dogpile_caching/caching_query.py. I'm guessing that https://github.com/zzzeek/sqlalchemy/blob/main/examples/dogpile_caching/caching_query.py#L140 already does that (by using the full statement + values as caching key) so it should already be providing that.

**Reply by zzzeek** [reply]
OK, caching query caches ORM results.

is the idea that you want to have ORM results from previous versions of your model in the cache?  that seems like it would be extremely difficult to achieve

**Reply by ddorian** [reply]
The idea was to not have clashes. But when you add/remove columns, the query will change, and it's being used as caching-key, so it will already invalidate by generating a new key. I don't need to add anything.
